rust/src/ssh/parser.rs

```diff
@@ -92,10 +112,91 @@ named!(pub ssh_parse_record<SshRecordHeader>,
         padding_len: be_u8 >>
         msg_code: be_u8 >>
         take!((pkt_len-2) as usize) >>
-        (SshRecordHeader{pkt_len, padding_len, msg_code})
+        (SshRecordHeader{pkt_len: pkt_len,
+            padding_len: padding_len,
+            msg_code: num::FromPrimitive::from_u8(msg_code).unwrap_or(MessageCode::SshMsgUndefined)})
+
     )
 );
 
+#[derive(Debug,PartialEq)]
+pub struct SshPacketKeyExchange<'a> {
+    pub cookie: &'a[u8],
+    pub kex_algs: &'a [u8],
+    pub server_host_key_algs: &'a [u8],
+    pub encr_algs_client_to_server: &'a [u8],
+    pub encr_algs_server_to_client: &'a [u8],
+    pub mac_algs_client_to_server: &'a [u8],
+    pub mac_algs_server_to_client: &'a [u8],
+    pub comp_algs_client_to_server: &'a [u8],
+    pub comp_algs_server_to_client: &'a [u8],
+    pub langs_client_to_server: &'a [u8],
+    pub langs_server_to_client: &'a [u8],
+    pub first_kex_packet_follows: u8,
+}
+
+use md5::compute;
```

Comment:
Why is it md5 ? Is it a standard ?

Reply:
No, but it's the most lightweight and best-fitting library.
I checked a few other options (https://lib.rs/keywords/md5)
Good options are: 
https://lib.rs/crates/md-5
https://lib.rs/crates/rust-crypto
I will make some performance tests.

Reply:
This comes in answer to other remarks as well
But did you know about md5 transform ?
https://suricata.readthedocs.io/en/suricata-5.0.3/rules/transforms.html#to-md5

Reply:
yes, I looked at it, but I was trying to make hassh functionality very similar to ja3/ja3s, because it's very similar 
https://suricata.readthedocs.io/en/suricata-5.0.0/rules/ja3-keywords.html
https://suricata.readthedocs.io/en/suricata-5.0.0/lua/lua-functions.html?highlight=ja3#ja3

I can use function like ```char *Ja3GenerateHash(JA3Buffer *buffer)``` to get hash.

Reply:
> Why is it md5 ? Is it a standard ?

It seems to be "HASSH is a network fingerprinting standard which can be used to identify specific Client and Server SSH implementations. The fingerprints can be easily stored, searched and shared in the form of a small MD5 fingerprint." (https://github.com/salesforce/hassh)

Reply:
Thanks @victorjulien 
I gather it is good to use `ssh.hassh` as a keyword as it is readable.
Then, should we somehow internally translate `ssh.hassh` to some `ssh.algorithms; to_md5` ?

Reply:
I think we can't really avoid having the md5 done directly as we should also consider the output case. In logging we'd also want to log the fingerprint (just like with ja3(s)).

Side thought: it would be interesting to see some benchmarks of different md5/sha1/sha256 implementations. How do the rust ones compare to libnss?

Reply:
So, thanks @victorjuilen and thanks @MalakhatkoVadym

Remaining questions is wether to use rust md5 crate or libnss md5 if I understand, right ?

Reply:
@catenacyber @victorjulien 
I created a function for benchmarking:
```
char* GenerateHash(const unsigned char* buffer, int len) {
    char *ja3_hash = SCMalloc(MD5_STRING_LENGTH * sizeof(char));
    if (ja3_hash == NULL) {
        SCLogError(SC_ERR_MEM_ALLOC,
                   "Error allocating memory for JA3 hash");
        return NULL;
    }

    unsigned char md5[MD5_LENGTH];
    HASH_HashBuf(HASH_AlgMD5, md5, (unsigned char *)buffer, len);

    int i, x;
    for (i = 0, x = 0; x < MD5_LENGTH; x++) {
        i += snprintf(ja3_hash + i, MD5_STRING_LENGTH - i, "%02x", md5[x]);
    }
    return ja3_hash;
}
```
rust md5 is : 9606 nanosec
libnsstime is : 18699 nanosec
rust md5 is : 10682 nanosec
libnsstime is : 16428 nanosec
rust md5 is : 4378 nanosec
libnsstime is : 7393 nanosec
rust md5 is : 4627 nanosec
libnsstime is : 7426 nanosec
rust md5 is : 10845 nanosec
libnsstime is : 15630 nanosec
rust md5 is : 10076 nanosec
libnsstime is : 16080 nanosec

Average time for rust md5 library is better than calling libnss function, so I'll leave md5 `compute` function.